Client/Frontend/Browser/BraveSearchHelper.swift

```diff
@@ -0,0 +1,116 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import Foundation
+import WebKit
+import Shared
+import BraveShared
+
+private let log = Logger.browserLogger
+
+class BraveSearchHelper: TabContentScript {
+    private weak var tab: Tab?
+    private let profile: Profile
+    
+    /// Tracks how many in current browsing session the user has been prompted to set Brave Search as a default
+    /// while on one of Brave Search websites.
+    private static var canSetAsDefaultCounter = 0
+    /// How many times user is shown the default browser prompt on Brave Search websites.
+    private let maxCountOfDefaultBrowserPromptsPerSession = 3
+    /// How many times user is shown the default browser prompt in total, this does not reset between app launches.
+    private let maxCountOfDefaultBrowserPromptsTotal = 10
+    
+    required init(tab: Tab, profile: Profile) {
+        self.tab = tab
+        self.profile = profile
+    }
+    
+    static func name() -> String { "BraveSearchHelper" }
+    
+    func scriptMessageHandlerName() -> String? { BraveSearchHelper.name() }
+    
+    private enum Method: Int {
+        case canSetBraveSearchAsDefault = 1
+        case setBraveSearchDefault = 2
+    }
+    
+    func userContentController(_ userContentController: WKUserContentController,
+                               didReceiveScriptMessage message: WKScriptMessage) {
+        var allowedHosts = ["search.brave.com",
+                            "search-dev.brave.com",
+                            "search-dev-local.brave.com",
+                            "search.brave.software",
+                            "search.bravesoftware.com"]
+        if !AppConstants.buildChannel.isPublic {
+            // TODO: Remove before merge.
+            allowedHosts.append("webpiaskownica.000webhostapp.com")
+        }
+        
+        guard let requestHost = message.frameInfo.request.url?.host,
+              allowedHosts.contains(requestHost),
+              message.frameInfo.isMainFrame else {
+            log.error("Backup search request called from disallowed host")
+            return
+        }
+        
+        guard let method = (message.body as? [String: Any])?["method_id"] as? Int else {
```

Comment:
Nit: Can we create Codable struct ( for instance MethodReference) here? And have method enum inside of it? It will be clearer to read.

Reply:
I didn't want to do it for 1 single property, can update it